### PR TITLE
easyeffects: add global settings and paired presets

### PR DESCRIPTION
**Pull request by Tenshock**
### Description

Add `services.easyeffects.settings` for declaratively managing global EasyEffects 8 settings.

Settings are grouped by KConfig section and written to EasyEffects' mutable database during Home Manager activation using `kwriteconfig6`. Boolean, integer, float, string, and path values are supported. Setting a value to `null` deletes the corresponding key.

Using activation instead of `xdg.configFile` keeps the database writable by EasyEffects. An assertion rejects non-empty settings when using EasyEffects versions older than 8.0.0.

Also allow one `services.easyeffects.extraPresets.<name>` entry to contain both `input` and `output` blocks. Each block is generated as a separate preset file in its corresponding directory:

```text
$XDG_DATA_HOME/easyeffects/input/<name>.json
$XDG_DATA_HOME/easyeffects/output/<name>.json
```

Existing one-sided input or output presets remain compatible.

Finally, extend `services.easyeffects.preset` with a structured form for selecting different startup presets for the input and output pipelines:

```nix
services.easyeffects.preset = {
  input = "voice";
  output = "music";
};
```

The existing string form remains supported. The structured form requires EasyEffects 8.0.0 or later and must select at least one pipeline; an empty string leaves that pipeline unchanged. After the service starts, Home Manager sends the pipeline-specific load commands to EasyEffects' local service socket.

Tests cover:

- String and boolean global settings.
- Key deletion using `null`.
- Generated activation commands.
- Unsupported EasyEffects version assertions.
- One-sided and paired input/output preset generation with exact JSON output.
- Backwards-compatible string startup presets.
- Pipeline-specific startup preset commands.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -A dev --run treefmt`.

- [x] Code tested through `nix build .#test-all`
      and targeted `nix run .#tests -- easyeffects`.

- [x] Test cases updated/added.

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

### CODE REVIEW

File: modules/services/easyeffects.nix [hunk 1]
```diff
@@ -9,12 +9,102 @@ let
...
+          ${splitPresetCommands}
+        } | ${pkgs.socat}/bin/socat -u -T 1 - UNIX-CONNECT:"$preset_socket"; then
+          exit 0
+        fi
+      fi
+
+      ${pkgs.coreutils}/bin/sleep 0.1
+    done
+
+    exit 1
```
Comment (hauskens) [inline modules/services/easyeffects.nix:65]: Normally i would prefer scripts to exit with a proper exit code as you do here, but in this case, the entire function is not essential for easyeffects to work, meaning it would potentially go in a restart loop if it fails due to `Restart = "on-failure";`. Should we maybe change this to `exit 0`, what are your thoughts on that?

Comment (Tenshock) [reply modules/services/easyeffects.nix:65]: Mmh good catch

I agree that failing to load the presets should not restart an otherwise working EasyEffects service.

What about prefixing ExecStartPost with `-`, as follows:
```diff
- ExecStartPost = "${loadSplitPresets} %t/EasyEffectsServer";
+ ExecStartPost = "-${loadSplitPresets} %t/EasyEffectsServer";
```

This makes the operation non-fatal to the unit while preserving the non-zero status and journal diagnostics ([prefix doc here](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Command%20lines)).

So we have best of both worlds, what do you think about this?

Comment (hauskens) [reply modules/services/easyeffects.nix:65]: That's even better, i was not aware you could even prefix commands like that in systemd units, so now i learned something new

Comment (Tenshock) [reply modules/services/easyeffects.nix:65]: I wasn't either, I also learned it just now while trying to find a better design!
I amended the last commit, we're all set

**Review by hauskens — APPROVED** [review]
Tested on my computer, works great!

File: modules/services/easyeffects.nix [hunk 2]
```diff
@@ -83,34 +182,85 @@ in
...
   config = lib.mkIf cfg.enable {
     assertions = [
       (lib.hm.assertions.assertPlatform "services.easyeffects" pkgs lib.platforms.linux)
+      {
+        assertion = !hasSettings || !olderThan8;
+        message = "${settingsOption} requires EasyEffects 8.0.0 or later.";
+      }
+      {
+        assertion = !isSplitPreset || !olderThan8;
+        message = "Structured `services.easyeffects.preset` requires EasyEffects 8.0.0 or later.";
```
Comment (khaneliman) [inline modules/services/easyeffects.nix]: issue (blocking): gate structured presets on the actual protocol boundary

This accepts EasyEffects 8.0.0, but 8.0.0–8.0.6 require numeric pipeline IDs and 8.0.0–8.0.8 listen through Qt's server-name resolution rather than `%t/EasyEffectsServer`. The current `input`/`output` protocol plus runtime path first works together in 8.0.9. Earlier versions wait through the helper timeout and silently skip preset loading because `ExecStartPost` is failure-prefixed. Require >=8.0.9 or implement version-specific protocol/path handling, with 8.0.8/8.0.9 boundary tests.

File: modules/services/easyeffects.nix [hunk 3]
```diff
@@ -83,34 +182,85 @@ in
...
+      {
+        assertion = !isSplitPreset || selectedPresets != { };
+        message = "Structured `services.easyeffects.preset` must select at least one input or output preset.";
+      }
     ];
 
     home.packages = with pkgs; lib.optional olderThan8 at-spi2-core ++ [ cfg.package ]; # Only include if easyeffects version is below 8.0.0
 
+    home.activation.easyeffectsSettings = lib.mkIf hasSettings (
+      lib.hm.dag.entryAfter [ "writeBoundary" ] ''
```
Comment (khaneliman) [inline modules/services/easyeffects.nix]: issue (blocking): do not update KConfig beneath the running service

Activation writes EasyEffects' mutable KConfig database while the service may still hold old values. Upstream autosaves periodically and calls `saveAll()` during shutdown, so a later stop can overwrite the declared values before the next process reads them. Apply settings while the service is stopped, such as from `ExecStartPre`, and add a restart trigger derived from `cfg.settings` so changes take effect deterministically.

File: modules/services/easyeffects.nix [hunk 4]
```diff
@@ -83,34 +182,85 @@ in
...
-        let
-          folder = builtins.head (builtins.attrNames v);
-        in
-        lib.nameValuePair "easyeffects/${folder}/${k}.json" {
-          source = jsonFormat.generate "${folder}-${k}.json" v;
-        }
+      lib.concatMapAttrs (
+        name: preset:
+        lib.mapAttrs' (
+          pipeline: settings:
```
Comment (khaneliman) [inline modules/services/easyeffects.nix:260]: issue (blocking): validate the complete preset key set

`presetType` checks only `lib.head (lib.attrNames v)`, but this loop consumes every attribute as a pipeline. `{ input = ...; outpt = ...; }` therefore passes and writes an unsupported `outpt` directory, while an empty set can fail inside `lib.head`. Require a non-empty set whose every key is `input` or `output`, and add typo and empty-set tests.

**Review by khaneliman — CHANGES_REQUESTED** [review]

File: modules/services/easyeffects.nix (hunk 2 above)
Comment (Tenshock) [reply modules/services/easyeffects.nix]: fixed: Require <= 8.0.9 for structured presets

[v8.0.9](https://github.com/wwmm/easyeffects/releases/tag/v8.0.9) was released on december 2025, 8 months ago. I've chosen to only support this feature for 8.0.9 and above.

File: modules/services/easyeffects.nix (hunk 3 above)
Comment (Tenshock) [reply modules/services/easyeffects.nix]: fixed: settings applied from `ExecStartPre`

I also added an `X-Restart-Triggers` value derived from `cfg.settings`, so changing the declared settings deterministically restarts the service. The tests now verify the pre-start script, restart trigger, generated `kwriteconfig6` commands, and that activation no longer writes the KConfig database.

File: modules/services/easyeffects.nix (hunk 4 above)
Comment (Tenshock) [reply modules/services/easyeffects.nix:260]: fixed: `presetType` now requires a non-empty attribute set and validates every key as either `input` or `output`.

I also the regression tests covering the case you gave as example and an empty preset set.
